torch/utils/data/sampler.py

```diff
@@ -334,25 +335,18 @@ def __init__(
     def __iter__(self) -> Iterator[List[int]]:
         # Implemented based on the benchmarking in https://github.com/pytorch/pytorch/pull/76951
         if self.drop_last:
+            sampler_iter = iter(self.sampler)
+            # Create multiple references to the same iterator
+            args = [sampler_iter] * self.batch_size
```

Comment:
It's unclear to me that this is safe in all scenarios although on the surface it seems OK. What benefit does this give us though?

Reply:
It's my understanding that it creates the required references first and then passes them all to `zip`, meaning it doesn't create a reference for each loop? I found [this SO post](https://stackoverflow.com/questions/2233204/how-does-zipitersn-work-in-python#:~:text=iter()%20is%20an%20iterator%20over%20a) quite informative on it.

I don't have the level of understanding of the pytorch codebase as either of you will, but I'm happy to work with you to test more things if that would be helpful. I found that the original list comprehension method was taking time when looping over large batches of small data when benchmarking some code with a colleague (@SusannaGreen).